extensions/oplogPopulator/OplogPopulator.js

```diff
@@ -242,35 +245,59 @@ class OplogPopulator {
+        return semver.gte(this._mongoVersion, constants.mongodbVersionWithImmutablePipelines);
+    }
+
+    async _initializeConnectorsManager() {
+        return this._connectorsManager.initializeConnectors();
+    }
+
     /**
      * Sets the OplogPopulator
      * @returns {Promise|undefined} undefined
      * @throws {InternalError}
      */
     async setup() {
-       try {
-           this._loadOplogHelperClasses();
-           this._connectorsManager = new ConnectorsManager({
-               nbConnectors: this._config.numberOfConnectors,
-               database: this._database,
-               mongoUrl: this._mongoUrl,
-               oplogTopic: this._config.topic,
-               cronRule: this._config.connectorsUpdateCronRule,
-               prefix: this._config.prefix,
-               heartbeatIntervalMs: this._config.heartbeatIntervalMs,
-               kafkaConnectHost: this._config.kafkaConnectHost,
-               kafkaConnectPort: this._config.kafkaConnectPort,
-               metricsHandler: this._metricsHandler,
-               logger: this._logger,
+        try {
+            this._loadOplogHelperClasses();
+            // initialize mongo client
+            await this._setupMongoClient();
+            const allocationStrategy = this.initStrategy();
+            this._connectorsManager = new ConnectorsManager({
+                nbConnectors: this._config.numberOfConnectors,
+                maximumBucketsPerConnector: this._maximumBucketsPerConnector,
+                database: this._database,
+                mongoUrl: this._mongoUrl,
+                oplogTopic: this._config.topic,
+                cronRule: this._config.connectorsUpdateCronRule,
+                prefix: this._config.prefix,
+                heartbeatIntervalMs: this._config.heartbeatIntervalMs,
+                kafkaConnectHost: this._config.kafkaConnectHost,
+                kafkaConnectPort: this._config.kafkaConnectPort,
+                metricsHandler: this._metricsHandler,
+                allocationStrategy,
+                logger: this._logger,
             });
-            await this._connectorsManager.initializeConnectors();
+            await this._initializeConnectorsManager();
             this._allocator = new Allocator({
                 connectorsManager: this._connectorsManager,
                 metricsHandler: this._metricsHandler,
+                allocationStrategy,
                 logger: this._logger,
             });
-            // initialize mongo client
-            await this._setupMongoClient();
+            // For now, we always use the RetainBucketsDecorator
+            // so, we map the events from the classes
+            this._connectorsManager.on('connector-destroyed', connector =>
+                allocationStrategy.onConnectorDestroyed(connector));
+            this._allocator.on('bucket-removed', (bucket, connector) =>
+                allocationStrategy.onBucketRemoved(bucket, connector));
```

Comment:
nit: can we do this?

```suggestion
            this._allocator.on('bucket-removed', allocationStrategy.onBucketRemoved);
```

Reply:
No, it breaks the unit tests